python/cuml/common/doc_utils.py

```diff
@@ -0,0 +1,411 @@
+
+    Currently auto detected variables include:
+    - X
+    - y
+    - convert_dtype
+    - sample_weights
+    - return_sparse
+    - sparse_tol
+
+    Typical usage scenarios:
+
+    Examples
+    --------
+
+    # for a function that passes all dense parameters, no need to specify
+    # anything, and the decorator auto detects the parameters and defaults
+
+    @generate_docstring()
+    def fit(self, X, y, convert_dtype=True):
+
+    # for a function that takes X as dense or sparse
+
+    @generate_docstring(X='dense_sparse')
+    def fit(self, X, y, sample_weight=None):
+
+    # to specify return values
+
+    @generate_docstring(return_values={'name': 'preds',
+                                       'type': 'dense',
+                                       'description': 'Predicted values',
+                                       'shape': '(n_samples, 1)'})
+
+
+    Parameters
+    -----------
+    X : str (default = 'dense')
+        Data type of variable X. Currently accepted types are: dense,
+        dense_anydtype, dense_intdtype, sparse, dense_sparse
+    X_shape : str (default = '(n_samples, n_features)')
+        Shape of variable X
+    y : str (default = 'dense')
+        Data type of variable y. Currently accepted types are: dense,
+        dense_anydtype, dense_intdtype, sparse, dense_sparse
+    y_shape : str (default = '(n_samples, 1)')
+        Shape of variable y
+    convert_dtype_cast : Boolean or str (default = False)
+        If not false, use it to specify when convert_dtype is used to convert
+        to a single specific dtype (as opposed to converting the dtype of one
+        variable to the dtype of another for example). Example of this is how
+        NearestNeighbors and UMAP use convert_dtype to convert inputs to
+        np.float32.
+    skip_parameters : list of str (default = [])
+        Use if you want the decorator to skip generating a docstring entry
+        for a specific parameter
+    skip_parameters_heading : boolean (deafault = False)
+        Set to True to not generate the Parameters section heading
+    return_values : dict or list of dicts (default = False)
+        Use to generate docstrings of return values. One dictionary per
+        return value, this is the format:
+            {'name': 'name_of_variable',
+             'type': 'data type of returned value',
+             'description': 'Description of variable',
+             'shape': 'shape of returned variable'}
+
+        If type is one of dense or dense_sparse then the type is generated
+        from the corresponding entry in _return_values_docstrings. Otherwise
+        the type is used as specified.
+    """
+
+    def deco(func):
+        params = signature(func).parameters
+
+        # Add parameter section header if needed, can be skipped
+        if(('X' in params or 'y' in params or parameters) and not
+                skip_parameters_heading):
+
+            func.__doc__ += \
+                '\nParameters\n----------\n'
```

Comment:
Was thinking about it and wanted to avoid having to look for the heading in the string for every occurrence of the decorator in the __doc__ string. So I left it to dev control with the `skip_parameters_heading`, particularly since the (pretty big) majority of use cases don't skip generating the heading, what do you think?